src/NuGet.Core/NuGet.DependencyResolver.Core/Remote/RemoteDependencyWalker.cs

```diff
@@ -107,15 +106,8 @@ public async Task<GraphNode<RemoteResolveResult>> WalkAsync(LibraryRange library
                     else
                     {
                         // Otherwise it's a dependency of this node
-                        if (dependencies == null)
-                        {
-                            // Init dependency lists
-                            dependencies = new List<LibraryDependency>(1);
-                            runtimeDependencies = new HashSet<string>();
-                        }
-
-                        dependencies.Add(libraryDependency);
-                        runtimeDependencies.Add(libraryDependency.Name);
+                        runtimeDependencies ??= new HashSet<LibraryDependency>(LibraryDependencyNameComparer.OrdinalIgnoreCaseNameComparer);
```

Comment:
How about using the List instead of HashSet here. I don't see any call to HashSet.Contains in this logic. If that works we can eliminate the call to `.ToList()` while creating `RemotResolveResult` object here https://github.com/NuGet/NuGet.Client/pull/5218/files#diff-7457360b2ac43e8123a03561b1b5e2ca1832954941f84e44086365d26f22e5faR147

Reply:
list allows duplicates, HashSet does not.

Reply:
Since this is expected to always have very few elements, let's swap to use a list for now. The original code had both, so this is a strict improvement already.

Reply:
I am unresolving this conversation because the current changes proposed has `runtimeDependencies.Contains` call twice. As we it is an O(n) operation. Looking at the call stack in https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1824305 and https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1824313, I think the best thing to do is first merge the changes to `CollectionExtensions.cs` in a separate PR and then we will just focus on `CreateGraphNode` method in this PR. What do you think?

Reply:
@kartheekp-ms, I originally made the change based on the suggestions from you and @zivkan. Based on feedback from @nkolev92, it seems that we can't rely on this collection always being small.

I'm inclined to revert the most recent change and go back to the implementation using `HashSet<LibraryDependency>`. We keep the O(1) check to avoid duplicates and the `ToList()` call will create a precisely sized list, so we'd save potential List resizes. Ideally, we'd just use the `HashSet<LibraryDependency>` directly as @nkolev92 suggests, but that would cause changes in several other areas that I'm not sure we'd want to take.

Let me know what you think.

Reply:
> I'm inclined to revert the most recent change and go back to the implementation using HashSet<LibraryDependency>. We keep the O(1) check to avoid duplicates and the ToList() call will create a precisely sized list, so we'd save potential List resizes.

Sorry for the confusion. Let's revert back to the previous commit.

Reply:
Reverted back. Let me know what you think.